Control dependency computation fix

danpoe opened the pull request:
Fixes a bug in the control dependency computation of the dependency graph (see regression test goto-instrument/dependence-graph4)

hannes-steffenhagen-diffblue commented:
Can you explain in more detail what you are doing here? What bug are you fixing here? Also, this seems to be a fairly significant change in the code - is this also a significant change in behavior, or did you combine your bug fix with a refactoring?

danpoe commented:
The PR does three things:

**(1)**

It fixes a bug in the control dependency computation. Previously in `transform()` only the control dependencies in the set `control_deps` (plus gotos and assumes at the `from` location) were considered candidates for the control dependencies at the `to` location. With this scheme however one misses control dependencies when there are nested if statements.

We use the following definition of control dependencies: A node N is control dependent on a node M iff N postdominates one but not all of M's CFG successors.

Consider the following example:
```
 1 void main(void)
 2 {
 3   if (a)
 4   {
 5     if (b)
 6       c=0; // no cd on if(a)
 7     else
 8       c=1; // no cd on if(a)
 9
10     d = 1; // cd on if(a)
11   }
12 }
```
Lines 6 and 8 do not have a control dependency on line 3, line 10 does have a control dependency on line 3. However, the previous implementation missed the latter dependency. The reason is that when computing the control dependencies at a node (e.g., line 10), only the control dependencies at the CFG predecessors (in this case lines 6 and 8) were considered candidates for the control dependencies.

This PR introduces a new set `control_dep_candidates` in addition to `control_deps`, which contains all control statements encountered so far.

**(2)**

It fixes an issue where `merge()` did not return true in a case where it should. The dependency graph has special handling for function entry edges (in `transform()`) where it merges the current state into the state of the return location before entering the function. This may change the abstract state at the return location. When later the function exit edge is handled, the `merge()` into the return location may not change the state, thus `merge()` may not return true, and thus the return location may not be inserted into the worklist of the fixpoint computation when it should.

This PR introduces a flag `has_changed` that records whether the abstract state has changed on the handling of the function entry edge. On the function return edge later, `merge()` checks the flag to know whether it needs to return true.

**(3)**

It simplifies the merging of the abstract states.

thk123 commented:
It would be helpful to have these as separate commits (or even PRs) to make it easier to review. 

For 3), does this PR help to address that/make it worse: #1843 In any case, could you have use `goto_programt::get_function_id`

thk123 commented on regression/goto-analyzer/dependence-graph4/test.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+--show --dependence-graph --text -
```
What's with the dash at the end?

danpoe replied:
`--text` takes an argument which is the file to print to, `-` means print to `stdout`.

thk123 commented on regression/goto-analyzer/dependence-graph4/test.desc:
```diff
@@ -0,0 +1,9 @@
+activate-multi-line-match
+EXIT=0
+SIGNAL=0
+Control dependencies: .\n\s*\n.*\n\s*g_out = 1;
```
I've not looked a the output of `--dependence-graph`, but what does the presence of this line signify? Should you be checking `t=1` is not present? 

Consider converting this to a unit test so you can be more precise.

danpoe replied:
It checks that there is a control dependency (the previous version wrongly reported no control dependency). It does however not check that the control dependency is to the correct instruction. This seems not to be possible easily in regression tests as control dependencies are reported as location numbers which are not stable across different systems.

I'll add a unit test for this.

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -26,36 +26,20 @@ bool dep_graph_domaint::merge(
 
-  depst::iterator it=control_deps.begin();
-  for(const auto &c_dep : src.control_deps)
+  if((has_values.is_false() || has_changed) &&
```
has_changed already includes the case that `has_values.is_false()`? I this is just `has_changed && data_deps.empty()`

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -114,14 +107,46 @@ void dep_graph_domaint::control_dependencies(
 }
 
+bool dep_graph_domaint::merge_control_dependencies(
+  const depst &other_control_deps,
```
To reduce the cognitive load, it would be great if all `deps` could be renamed to `dependencies`

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -114,14 +107,46 @@ void dep_graph_domaint::control_dependencies(
+
+  // Merge control dependencies
+
+  size_t n=control_deps.size();
```
Perhaps `n` could be `original_control_deps_size`

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -114,14 +107,46 @@ void dep_graph_domaint::control_dependencies(
+
+  // Merge control dependency candidates
+
+  n=control_dep_candidates.size();
```
And this `n` could be `original_control_dep_candidate_size`

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -189,37 +214,42 @@ void dep_graph_domaint::transform(
   ai_baset &ai,
   const namespacet &ns)
 {
+  INVARIANT(!has_values.is_false(),
```
`PRECONDITION` probably makes more sense here

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -189,37 +214,42 @@ void dep_graph_domaint::transform(
   if(from->is_function_call())
   {
-    if(from->function == to->function)
+    goto_programt::const_targett next=from;
```
`const goto_programt::const_targett next=std::next(from);`

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -189,37 +214,42 @@ void dep_graph_domaint::transform(
-
       dep_graph_domaint *s=
         dynamic_cast<dep_graph_domaint*>(&(dep_graph->get_state(next)));
-      assert(s!=nullptr);
```
Why is this assert no longer required?

thk123 reviewed (commented):
I don't fully understand the definition of the control dependency so going to pause reviewing. How come in your example line 6 doesn't have control dependency on line 3? My reading of your definition led me to, 

3 -> 5 -> 6 -> 10 (postdominates one)
3 -> 5 -> 7 -> 8 -> 10 (doesn't postdominate them all)

martin-cs commented on src/analyses/dependence_graph.cpp:
```diff
@@ -67,42 +51,51 @@ void dep_graph_domaint::control_dependencies(
-  if(from->is_goto() ||
-     from->is_assume())
-    control_deps.insert(from);
+  // The "successor" above refers to an immediate successor of M.
```
Why is the assume case omitted?

martin-cs commented on src/analyses/dependence_graph.cpp:
```diff
@@ -67,42 +51,51 @@ void dep_graph_domaint::control_dependencies(
-      pd.cfg.entry_map.find(*it);
-
-    assert(e!=pd.cfg.entry_map.end());
+      pd.cfg.entry_map.find(cd);
```
Is this assertion no longer true?

martin-cs commented on src/goto-programs/goto_program_template.h:
```diff
@@ -386,6 +386,14 @@ class goto_program_templatet
+
+    return l->function;
+  }
+
```
Didn't we just remove this method?

danpoe replied:
Indeed 5df3fca9cfe7b9732f0d87d1a0eaf740194bab87 removes this method. It seems that since by f443b1815e34bbec2b51fe14d732ba41d92010b7 cbmc does not do partial inlining anymore, the function field of an instruction is now guaranteed to contain the id of the function it is in. I'll remove the method again.

(In goto-instrument there are still a few cases where the function field is not adjusted when the inliner is called explicitely, but we can fix this in the future.)

martin-cs commented:
I'm tempted to be lazy and say +1 "what @thk123" said.  Separate commits would be great.  The brilliant explanations here : https://github.com/diffblue/cbmc/pull/1858#issuecomment-366731732 as commit messages would be even better.

danpoe commented:
@thk123 Line 6 does not have a control dependency on line 3 as it does not postdominate a CFG successor of line 3 (i.e., `postdominates_one` is false). The two CFG successors of line 3 are line 5 and line 11. Starting from line 5, line 6 does not occur on all paths to the exit point. Similarly, starting from line 11, line 6 also does not occur on all paths to the exit point. Thus, line 6 does not postdominate a CFG successor of line 3.

danpoe commented:
I'll split the PR up into several commits.

danpoe committed:
Simplified state merging in the dependence graph

danpoe committed:
Control dependency computation fix

Previously in transform() only the control dependencies in the set control_deps
(plus gotos and assumes at the from location) were considered candidates for the
control dependencies at the to location. With this scheme however one misses
control dependencies when there are nested if statements.

We introduce a new set control_dep_candidates in addition to control_deps, which
contains all control statements encountered so far.

danpoe committed:
Make dependence graph merge() return true when abstract state changes

The dependency graph has special handling for function entry edges (in
transform()) where it merges the current state into the state of the return
location before entering the function. This may change the abstract state at the
return location. When later the function exit edge is handled, the merge() into
the return location may not change the state, thus merge() may not return true,
and thus the return location may not be inserted into the worklist of the
fixpoint computation when it should.

We introduce a flag has_changed that records whether the abstract state has
changed on the handling of the function entry edge. On the function return edge
later, merge() checks the flag to know whether it needs to return true.

smowton commented on src/analyses/dependence_graph.cpp:
```diff
@@ -26,37 +26,23 @@ bool dep_graph_domaint::merge(
-    }
-    else if(it!=data_deps.end())
-      ++it;
+    data_deps = src.data_deps;
```
Explain why when our own data deps are non-empty, we don't use src's data deps at all?

danpoe replied:
We wouldn't actually need a fixpoint computation for the data dependency analysis as it merely uses the result of the prior reaching definitions analysis to compute the data dependencies at each location. For this we need to inspect each location only once. Thus if the set `data_deps` is non-empty it means we have handled that location already.

That's one reason why it may be good to split up the data dependency and control dependency analysis (see issue #1859), and then implement the data dependency analysis without `ait` by simply iterating over all instructions.

smowton commented on src/analyses/dependence_graph.cpp:
```diff
@@ -67,42 +53,54 @@ void dep_graph_domaint::control_dependencies(
+  // The "successor" above refers to an immediate successor of M.
+
+  // Candidates for M for "to" are "from" and all existing control
+  // dependencies on nodes. "from" is added if it is a goto or assume
```
"all existing control dependencies on nodes" is less clear than the original phrasing I think. How about "all nodes X such that N is control-dependent on X"

danpoe replied:
I just realized that this comment is not accurate anymore in the fixed version. Now we consider all control statements in the current function from which there is a path to the current statement (i.e., `to`) as potential sources of control dependency edges. I'll add a comment to explain this.

smowton commented on src/analyses/dependence_graph.cpp:
```diff
@@ -67,42 +53,54 @@ void dep_graph_domaint::control_dependencies(
+  // Check all candidates
 
-    // check all CFG successors
+  for(const auto &cd : control_dep_candidates)
```
I think this bit is just cleanup for clarity and not a substantial change? If so, if not already, move it into a different commit than the substantial change.

danpoe replied:
The set `control_dep_candidates` was introduced in the fix so this is new code.

smowton commented on src/analyses/dependence_graph.cpp:
```diff
@@ -114,14 +112,43 @@ void dep_graph_domaint::control_dependencies(
+
+  size_t n = control_deps.size();
+
+  control_deps.insert(other_control_deps.begin(), other_control_deps.end());
```
It seems a shame to degrade the previous more-efficient algorithm that makes a single walk over the target set into this one that makes |other_control_deps| searches from the root. If it bothers you that the old algorithm is less clear, how about factoring it into a util called `union_ordered_range` or something?

danpoe replied:
Indeed. I assumed that it would implement the same algorithm if both ranges are sorted but it's not the case. I'll add a `union_ordered_range` utility function.

smowton reviewed (commented):
Main problem here: this absolutely cries out for a reasonably-thorough batch of test cases, since this is the 3rd slicer correctness fix to be PR'd in the last 6 months.

danpoe commented:
@smowton I'll add some more dependence graph tests.

danpoe committed:
More dependence graph tests

danpoe commented:
@smowton, @thk123, @martin-cs: CI is passing and I've addressed all the comments. I think this is now ready to be merged.

thk123 commented on src/util/container_utils.h:
```diff
@@ -0,0 +1,40 @@
+#define CPROVER_UTIL_CONTAINER_UTILS_H
+
+template <class C>
+bool union_ordered_range(C &target, const C &source)
```
Document please

thk123 replied:
In particular any requirements on what C can be (e.g. iterable, ordered?)

thk123 commented on src/analyses/dependence_graph.h:
```diff
@@ -162,7 +162,9 @@ class dep_graph_domaint:public ai_domain_baset
   typedef std::set<goto_programt::const_targett> depst;
-  depst control_deps, data_deps;
+  depst control_deps;
+  depst control_dep_candidates;
```
It would be good to document here what the difference between a control dependency and a control dependency candidate is

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -48,42 +50,54 @@ void dep_graph_domaint::control_dependencies(
+  // Check all candidates
 
-    // check all CFG successors
+  for(const auto &cd : control_dep_candidates)
```
Expand `cd` (and possibly call it `control_dep_candidate` since it is a candidate not a control dependency(?? - I don't really know the difference)

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -48,42 +50,54 @@ void dep_graph_domaint::control_dependencies(
 
-    assert(e!=pd.cfg.entry_map.end());
+    INVARIANT(
+      e != pd.cfg.entry_map.end(), "cfg has an entry for every location");
```
The wording for this invariant is a little confusing - do you mean cfg _should_ have an entry for every location?

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -27,19 +27,20 @@ bool dep_graph_domaint::merge(
   goto_programt::const_targett from,
```
Nit: this commit message (Fix issue where merge() did not return true in the dependence graph w) isn't rendering correctly so I suggest the title is too long. Suggest making the short text:

```
Make dependency graph merge() return true when abstract state changes
```

I think that is short enough

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -27,19 +27,20 @@ bool dep_graph_domaint::merge(
+  if(is_bottom() || (has_changed && data_deps.empty()))
   {
-    has_values = tvt::unknown();
     data_deps = src.data_deps;
```
Suggest something like the above comment goes into the code as this is entirely non-obvious! 

It also might be clear to do:

```
if(changed)
{
  // ...
  // comment explaining why not if data_deps in non-empty
  if(!data_deps.empty())
  {
    data_deps = src.data_deps;
  }
}
```

thk123 reviewed (commented):
Changes and commits look good.
Some requests for docs 
I am not familiar enough with this to work out if the tests are correct - though it would be easier to check them if there was a comment at the bottom of the `desc` decoding the regex to see what _is_ expected

danpoe commented:
@thk123 By adding a comment to the desc file which decodes the regex, do you mean adding an example (possibly with some placeholders for e.g. location numbers) for the portion of the output that the regex is designed to match?

thk123 commented:
Comments can be added after two sets of `--` (sorry if you already knew that). Then my suggested comment would be:

```
CORE
main.c
--dependence-graph --show
activate-multi-line-match
EXIT=0
SIGNAL=0
// First assignment has a control dependency on the if
\/\/ ([0-9]+).*\n.*IF.*i < 10.*THEN(.*\n)*Control dependencies: \1\n(.*\n){2,3}.*a = 1
// Second assignment has a control dependency on the if
\/\/ ([0-9]+).*\n.*IF.*i < 10.*THEN(.*\n)*Control dependencies: \1\n(.*\n){2,3}.*a = 2
--
^warning: ignoring
--
Matches:
// IF i < 10 THEN
// Control dependencies: 1
// a = 1
and
// IF i < 10 THEN 
// Control dependencies: 1
// a = 2
```

Assuming of course I've decoded your regex to approx right thing

thk123 commented on src/analyses/dependence_graph.cpp:
```diff
@@ -67,42 +58,54 @@ void dep_graph_domaint::control_dependencies(
 
-    assert(e!=pd.cfg.entry_map.end());
+    INVARIANT(
+      e != pd.cfg.entry_map.end(), "cfg has an entry for every location");
```
The wording here looks the same, I suggest "cfg must have an entry for every location"

danpoe replied:
Ah yes. I thought I'd change it once #2142 is resolved, but I'll rephrase it now with must.

thk123 commented on regression/goto-analyzer/dependence-graph10/test.desc:
```diff
@@ -0,0 +1,25 @@
+// First assignment has a control dependency on the if
+\/\/ ([0-9]+).*\n.*IF.*i < 10.*THEN(.*\n)*Control dependencies: \1\n(.*\n){2,3}.*a = 1
+// Second assignment has a control dependency on the if
+\/\/ ([0-9]+).*\n.*IF.*i < 10.*THEN(.*\n)*Control dependencies: \1\n(.*\n){2,3}.*a = 2
```
There's only one `IF i < 10` how come it appears here twice with both dependencies?

danpoe replied:
That's because `a = 1` and `a = 2` appear in the then and else branches of this if. So the statements in both branches have a control dependency on the goto that decides which branch to take.

thk123 commented on regression/goto-analyzer/dependence-graph10/test.desc:
```diff
@@ -0,0 +1,25 @@
+--
+^warning: ignoring
+--
+The two regexes above match output portions like shown below (with <N> being a
```
This explanation is great, though I was hoping to have both lines laid about explicitly

thk123 commented on regression/goto-analyzer/dependence-graph12/test.desc:
```diff
@@ -0,0 +1,23 @@
+EXIT=0
+SIGNAL=0
+// Assignment has a control dependency on the if
+\/\/ ([0-9]+).*\n.*IF.*a < 10.*THEN(.*\n)*Control dependencies: .*\1.*\n(.*\n){2,3}.*a = 2
```
Why don't we care what has a dependency on `a < 7`?

thk123 reviewed (approved):
Someone with a good understanding of control dependencies should check the tests!

danpoe commented:
Ok, I've had a chat with @marek-trtik who worked on the slicer recently. He's up for double checking the control dependency regression tests.

marek-trtik commented on regression/goto-analyzer/dependence-graph12/test.desc:
```diff
@@ -0,0 +1,38 @@
+\/\/ ([0-9]+).*\n.*IF.*a < 7.*THEN(.*\n)*Control dependencies: (([0-9]+,\1)|(\1,[0-9]+))\n(.*\n){2,3}.*a = 2
+// Assignment has a control dependency on the second if
+\/\/ ([0-9]+).*\n.*IF.*a < 10.*THEN(.*\n)*Control dependencies: (([0-9]+,\1)|(\1,[0-9]+))\n(.*\n){2,3}.*a = 2
+--
```
why checking only for `a=2` and not also for `a=1`?

danpoe replied:
Yes, I can add a check for `a = 1` too.

marek-trtik commented on regression/goto-analyzer/dependence-graph11/test.desc:
```diff
@@ -0,0 +1,23 @@
+location number). The intention is to check whether the assignment has a control
+dependency on the goto statement.
+
+        // <N> file main.c line 7 function main
```
line 11

danpoe replied:
The test regexes above actually ignore the line numbers and the comments currently just give examples of the kind of output that they would match. But I can change the line numbers in the output examples to match the tests.

marek-trtik commented on regression/goto-analyzer/dependence-graph11/test.desc:
```diff
@@ -0,0 +1,23 @@
+**** 3 file main.c line 9 function main
+Control dependencies: <N>
+
+        // 3 file main.c line 9 function main
```
line 13

marek-trtik commented on regression/goto-analyzer/dependence-graph4/test.desc:
```diff
@@ -0,0 +1,22 @@
+**** 3 file main.c line 9 function main
+Control dependencies: <N>
+
+        // 3 file main.c line 9 function main
```
line 14

marek-trtik commented on regression/goto-analyzer/dependence-graph4/test.desc:
```diff
@@ -0,0 +1,22 @@
+EXIT=0
+SIGNAL=0
+\/\/ ([0-9]+) file.*\n.*IF.*g_in1.*THEN GOTO(.*\n)*Control dependencies: \1\n\n.*\n.*g_out = 1
+--
```
Would it be possible to check that neither 10 nor 12 have CD on 7?

danpoe replied:
Yes, I'll add a check for that too.

marek-trtik commented on regression/goto-analyzer/dependence-graph4/main.c:
```diff
@@ -0,0 +1,16 @@
+int g_in1, g_in2;
+int g_out;
```
Why global vars and not automatic?

danpoe replied:
There is no reason for this. The variables do not matter for control dependencies.

marek-trtik reviewed (commented):
Half done, finish on Tuesday ...

marek-trtik commented on regression/goto-analyzer/dependence-graph7/test.desc:
```diff
@@ -0,0 +1,50 @@
+**** 4 file main.c line 6 function main
+Control dependencies: <N>
+
+        // 4 file main.c line 6 function main
```
Should not be here rather line 8 or 9?

danpoe replied:
It seems that the source location assigned to the back edge of a loop is the source location of the loop head.

marek-trtik commented on regression/goto-analyzer/dependence-graph8/test.desc:
```diff
@@ -0,0 +1,54 @@
+location number). The intention is to check whether the assignment has a control
+dependency on the if.
+
+        // <N> file main.c line 6 function main
```
line 8

marek-trtik commented on regression/goto-analyzer/dependence-graph8/test.desc:
```diff
@@ -0,0 +1,54 @@
+**** 3 file main.c line 8 function main
+Control dependencies: <N>
+
+        // 3 file main.c line 8 function main
```
line 10

marek-trtik commented on regression/goto-analyzer/dependence-graph7/test.desc:
```diff
@@ -0,0 +1,50 @@
+dependency on itself.
+
+Control dependencies: <N>
+Data dependencies: 1
```
We focus only on CD. So, this info seams to be redundant/misleading? Remove? What do you think?

danpoe replied:
The dependency graph outputs both the control dependencies and the data dependencies. The regex above ignores the data dependencies actually. But the example here shows the typical output that the dependency graph produces and which would be matched by the above regex.

marek-trtik reviewed (approved):
The regression tests look good to me.

smowton commented on src/util/container_utils.h:
```diff
@@ -0,0 +1,57 @@
+/// \param source second input set
+/// \return true iff `target` was changed
+template <class T, class Compare, class Alloc>
+bool util_set_union(
```
Suggest `_union_inplace`, since `set_union` would usually be a binary op that returns a new set.

tautschnig replied:
1) I'd suggest `inplace_set_union` as that would be in keeping with the STL's `inplace_merge`.
2) Any ideas of how much faster this implementation is over: 
````c++
std::size_t before = target.size();
decltype(target) tmp;
tmp.swap(target);
std::set_union(tmp.begin(), tmp.end(), source.begin(), source.end(), target.end());
return target.size() != before;
````
? Just curious.

danpoe replied:
I just tried this, with `std::set<std::size_t>`'s, both sets being of equal size, and the sets containing random values. The runtimes (in microseconds) for the unions are as follows (union1 is the version in the PR). The leftmost column gives the sizes of the two sets being merged.

```
       union1     union2
100        39         75
  1K      368        698
 10K    3,358      6,704
100K   54,455    104,597
  1M  592,993  1,208,903
```

tautschnig replied:
Thank you!! I suppose that memory allocation is the dominating factor, which of course has to be done from scratch when using `std::set_union`. So I'd still suggest to rename the function to some variation involving "inplace", and then this PR should just be merged?!

smowton reviewed (approved):
One nitpick, otherwise lgtm.

tautschnig reviewed (commented):
I am confused by the comment "With this scheme however one misses control dependencies when there are nested if statements." -- are you now computing the transitive closure?

danpoe commented:
We're not computing transitive dependencies. It's just that the bug that this PR fixes appeared when there was an if statement (with both an if and an else branch) within another if statement. Then for the code that follows the inner if statement (and is contained within the outer if statement), we would not report a control dependency on the outer if condition (as shown in the example above).

tautschnig commented:
> [...] we would not report a control dependency on the outer if condition [...]

It's not clear to me whether that should be reported? The control dependency on the outer if is a transitive one, and would be found by looking at the control dependency of the inner if. But maybe I'm misunderstanding.

danpoe commented:
In the example above, for line 10 currently no control dependency is reported (neither on line 3 nor on line 5). It's correct that it does not have a control dependency on line 5, but we need to report a control dependency on line 3. As no control dependency is currently reported for line 10, there's no dependency chain that we could follow to know that it depends on line 3.

tautschnig commented:
Thanks a lot for the clarification!
